### ISSUE TITLE
Logout of app on multi app tenant pings all apps

### ISSUE THREAD

**Issue by davidmw** [FusionAuth/fusionauth-issues#316]
### Description 
When the user logs out of one of our apps on 1.10.0 we are seeing requests made of all applications in the tenant.

/auth0/logout endpoint is what we redirect to on the user's logout request.  In the browser network tab I'm seeing requests made of APIs that are specific to each app which suggests that the logout process is iterating thru our apps.

An example cross-app request is:
`https://<app>.z8.web.core.windows.net/?iss=identtest.<domain>&sid=ECDE4xxxxC`

The logout flow captured in Network tab is:

```
https://accounts.google.com/o/oauth2/iframe
https://<appOne>.z8.web.core.windows.net/?iss=identtest.<domain>&sid=ECDE4xxxxC
https://<appTwo>.z8.web.core.windows.net/?iss=identtest.<domain>&sid=ECDE4xxxxC (app in browser)
https://<appTwo>.z8.web.core.windows.net/?iss=identtest.<domain>&sid=ECDE4xxxxC
https://<appThree>.z8.web.core.windows.net/?iss=identtest.<domain>&sid=ECDE4xxxxC
https://identtest<domain>/oauth2/logout?redirect_uri=... (there are two redirect_uri on this request and two different client IDs)
/oauth2/authorize?
```

Our test server and your account remain available if you want to have a look for yourself.  Login to the same app as instructed for the Google sign in support issue.

Please note that this logout behaviour exists if the App is configured with a logout URL or not.

**Comment by davidmw** [comment]
This is looking like an issue with the  release notes:

> TL;DR The /oauth2/logout endpoint can call logout URLs of multiple applications and a specific Redirect URL can be requested on the URL via post_logout_redirect_uri.

This might be better worded as 'endpoint will call logout URLs of all tenant applications'

**Comment by robotdan** [comment]
Thanks for the feedback @davidmw I'll update the release notes. 

Do you have a use case where you do not want to log out of all applications in the tenant?

**Comment by davidmw** [comment]
@robotdan We are multi tenant for release staging. In each is three SPAs and one common API collection as apps. The largest user group, students, will never be in two of the SPAs. Each SPA has 1 or 2 unique API collections behind it and these are serverless function apps on azure. 

When students logout it is pointless to attempt logouts of the other apps. Worse, attempts to logout will activate and cause autoscaling of multiple serverless apps that should not have been hit and could attempt to load dependencies from external domains that are not whitelisted for students. 

Using post_logout_redirect_uri cleans up the logout process but I’m finding that it results in an automatic load of /login for default tenant immediately after our redirect. So logout is causing some headaches right now.

**Comment by davidmw** [comment]
post_logout_redirect_uri isn't solving the problem for our second app. The redirect URL is in the authorised section of the app configuration. /oauth2/logout has post_logout_redirect_uri and client_id on the query string.  All apps in the tenant are pinged on logout.

In the first app I've been able to remove post_logout_redirect_uri and it still (correctly) only logs out of the intended app. If I have my {{logout url}} on post_logout_redirect_uri then something is causing that page to be bypassed and the browser ends up on /login. If {{logout url}} is instead in the App configuration then we stop on the correct {{logout url}}.

My {{logout url}} now always appears in the browser with a query string added to it like:

>  ?iss=identtest.{{domain}}&sid=B4AA...E91CB5

**Comment by robotdan** [comment]
Can you clarify the issue you're seeing, I'm trying to wade through the details. :-) 

You are redirecting the browser to `/oauth2/logout` and you do not want to logout of all applications within the tenant, but just the one for which you're providing the `client_id`? 

> In the first app I've been able to remove post_logout_redirect_uri and it still (correctly) only logs out of the intended app. If I have my {{logout url}} on post_logout_redirect_uri then something is causing that page to be bypassed and the browser ends up on /login. If {{logout url}} is instead in the App configuration then we stop on the correct {{logout url}}.

Not sure I follow, can you please clarify? Thanks.

**Comment by davidmw** [comment]
> You are redirecting the browser to /oauth2/logout and you do not want to logout of all applications within the tenant, but just the one for which you're providing the client_id?

Correct. When I opened this ticket the student app was causing logout of all apps within the tenant.  After experimenting with post_logout_redirect_uri this 'logout all' behaviour stopped but the browser was not staying on the URI specified in the redirect. 

In our second app I'm unable to disable the 'logout all' behaviour. I'm unsure what corrected this behaviour on the student app since including or omitting post_logout_redirect_uri is not having an effect.

**Comment by robotdan** [comment]
The intended behavior is to log out all applications in the tenant. If we don't do this, logging out of the SSO really isn't complete since. 

Adding the `post_logout_redirect_uri` should only affect the redirect URI, but not the behavior of the logout. 

So it sounds like you have a use case where you want to call the OAuth2 Logout endpoint (`/oauth2/logout`) and only log out of SSO and the app that called you and leave all other applications alone. 

This is not the design point, but perhaps it is a valid use case that we need to account for. 

A work around would be to remove all logout URL configuration, and only use the `post_logout_redirect_uri` parameter to perform the redirect where you could also perform logout of the application.

**Comment by davidmw** [comment]
> So it sounds like you have a use case where you want to call the OAuth2 Logout endpoint (/oauth2/logout) and only log out of SSO and the app that called you and leave all other applications alone

As a point of clarification, the design of logout is executing log out on apps not initialized during the session and that the user is typically not registered for and not authorized for.

**Comment by robotdan** [comment]
> As a point of clarification, the design of logout is executing log out on apps not initialized during the session and that the user is typically not registered for and not authorized for.



This is a good point @davidmw . The logout endpoints (in a general sense) should be idempotent, but I can appreciate the use case. 

I would hesitate to use registration / authorization during the logout workflow to limit the logout URLs because this may not reflect if the user has a session on those other services. For example, the user may have recently had their registration revoked but they may still have an active session. 

I would propose the following enhancement: 

Application configuration options:
- Logout of none (legacy behavior, which that we will end your SSO session and nothing else, redirect to the registered logout URL or the requested `post_logout_redirect_uri`)
- Logout of some, provide an additional request parameter `additionanl_client_id` and specify this parameter one to many times to indicate which applications you would like the Logout URL to be called
- Logout of all (current - new - behavior where the logout URL of all applications is called)


Would this better suite your requirements? Or do you have any feedback on this this solution? 

Does this work around I described above help you for now? 

> A work around would be to remove all logout URL configuration in each application and at the tenant configuration level, and only use the `post_logout_redirect_uri` parameter to perform the redirect where you could also perform logout of the application.

**Comment by davidmw** [comment]
Yes the workaround solves the problem for us, thank you.  And Logout of none, some, and all sounds like a great approach.

Much of our trouble seems due to differences in what we understood 'logout url' to be.  Our app logout process is complete before we call /oauth2/logout.  So we want to go from that to a post logout url which is typically the application root.

**Comment by robotdan** [comment]
Thanks for that additional detail @davidmw . 

> Yes the workaround solves the problem for us, thank you. And Logout of none, some, and all sounds like a great approach.


Great, we will deliver this in an upcoming release.
